cpp/src/arrow/util/value_parsing.h

```diff
@@ -561,6 +562,38 @@ static inline bool ParseTimestampStrptime(const char* buf, size_t length,
   return true;
 }
 
+/// \brief Returns time since the UNIX epoch in the requested unit. Takes null terminated
+/// buffer as argument
+static inline bool ParseTimestampStrptime(const char* buf, const char* format,
+                                          bool ignore_time_in_day,
+                                          bool allow_trailing_chars, TimeUnit::type unit,
+                                          int64_t* out) {
+  // NOTE: strptime() is more than 10x faster than arrow_vendored::date::parse().
+  struct tm result;
+  memset(&result, 0, sizeof(struct tm));
+#ifdef _WIN32
+  char* ret = arrow_strptime(buf, format, &result);
+#else
+  char* ret = strptime(buf, format, &result);
```

Comment:
can we put a switch here to do the copy only if buf[len-1] = '\0' ? That'll avoid duplication of code.

Reply:
There is still a copy in the constructor of clean_copy. I tried running the benchmark with this but the execution time doesn't improve much if we don't remove the constructor.

Reply:
Also c++ stl string permits '\0' as character, so we can't reject it based on that if method doesn't take null terminated string by design

Reply:
sounds good, you are right.